README.md
# [Nest Comment](http://bootstrap-table.wenzhixin.net.cn)

[![Build Status](https://travis-ci.org/wenzhixin/bootstrap-table.svg)](https://travis-ci.org/wenzhixin/bootstrap-table) 
[![GitHub version](https://badge.fury.io/gh/wenzhixin%2Fbootstrap-table.svg)](http://badge.fury.io/gh/wenzhixin%2Fbootstrap-table)
[![Bower version](https://badge.fury.io/bo/bootstrap-table.svg)](http://badge.fury.io/bo/bootstrap-table)
[![jsDelivr Hits](https://data.jsdelivr.com/v1/package/npm/bootstrap-table/badge?style=rounded)](https://www.jsdelivr.com/package/npm/bootstrap-table)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ZDHP676FQDUT6)

[![](https://graphs.waffle.io/wenzhixin/bootstrap-table/throughput.svg)](https://graphs.waffle.io/wenzhixin/bootstrap-table/throughput.svg)

An extended Bootstrap table with radio, checkbox, sort, pagination, extensions and other added features.

To get started, check out:

* [Docs](http://bootstrap-table.wenzhixin.net.cn)
* [Examples](https://github.com/wenzhixin/bootstrap-table-examples)
* [Questions/Helps](http://stackoverflow.com/questions/tagged/bootstrap-table)
* [问题/帮助](http://segmentfault.com/t/bootstrap-table)


[**List of donators**](https://github.com/wenzhixin/bootstrap-table/blob/master/DONATORS.md)

## LICENSE

**NOTE:** Bootstrap Table is licensed under the [The MIT License](https://github.com/wenzhixin/bootstrap-table/blob/master/LICENSE). Completely free, you can arbitrarily use and modify this plugin. If this plugin is useful to you, you can **Star** this repo, your support is my biggest motive force, thanks.

## Features

* Created for Bootstrap 3 (Bootstrap 2 supported)
* Responsive web design
* Scrollable Table with fixed headers
* Fully configurable
* Via data attributes
* Show/Hide columns
* Show/Hide headers
* Show/Hide footers
* Get data in JSON format using AJAX
* Simple column sorting with a click
* Format column
* Single or multiple row selection
* Powerful pagination
* Card view
* Detail view
* Localization
* Extensions

## How to get it

### Manual download

Use [Releases page](https://github.com/wenzhixin/bootstrap-table/releases) or [the source](https://github.com/wenzhixin/bootstrap-table/archive/master.zip).

### Bower

```
bower install bootstrap-table
```

### Npm

```
npm install bootstrap-table
```

### CDN

You can source bootstrap-table directly from a CDN like [CDNJS](http://www.cdnjs.com/libraries/bootstrap-table) or [bootcss](http://open.bootcss.com/bootstrap-table/) or [jsdelivr](http://www.jsdelivr.com/#!bootstrap.table).


## Contributing

For feature requests, bug reports or submitting pull requests, please ensure you first read [CONTRIBUTING.md](https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md).


## Reporting Issues

As stated above, please read [CONTRIBUTING.md](https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md), especially [Bug Reports](https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#bug-reports)

And as stated there, please provide jsFiddle when creating issues!

It's really saves much time.

You can also use these for templates:

[jsFiddle Bootstrap Table](http://bootstrap-table.wenzhixin.net.cn/examples/#basic)

Your feedback is very appreciated!


## Acknowledgements

Thanks to everyone who have given feedback and submitted pull requests. A list of all the contributors can be found [here](https://github.com/wenzhixin/bootstrap-table/graphs/contributors).

## Release History

Look at the [Change Log](https://github.com/wenzhixin/bootstrap-table/blob/master/CHANGELOG.md)

## Local build

To build bootstrap-table locally please run:

```
npm run grunt
```

Result will appear in `dist` directory.


CHANGELOG.md
ChangeLog
---------

### 1.0.0

- Initial release


index.js
/**
 * Created by liuqi on 2017/12/11.
 */
$(function () {
    /*$('.comment-box').nestComment({
     data:[{
     id: 1,
     name: 'Item 1',
     price: '$1'
     }, {
     id: 2,
     name: 'Item 2',
     price: '$2'
     }]
     });*/

    /*$('#my-nest').nestComment({
     url: '/nest/comments?parentId=0&topicType=0&topicId=24695531230177356545&type=0&page=1&rows=10',
     onAll:function (data,args) {
     console.log('双击666');
     }
     });*/
    $('#my-nest').nestComment({
        data: [{
            "id": "24719563439646783560",
            "authorName": "jelon",
            "authorEmail": "jilijelon@163.com",
            "authorUrl": "www.baidu.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "firefox",
            "authorInfo": null,
            "createDate": "2017-12-07 14:59:44",
            "content": "不错哦，双击666",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 1,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24719566450206817150",
            "authorName": "joanna",
            "authorEmail": "joanna@163.com",
            "authorUrl": "www.baidu.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "firefox",
            "authorInfo": null,
            "createDate": "2017-12-07 15:00:02",
            "content": "棒棒棒",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 1,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24719567714396197995",
            "authorName": "joanna",
            "authorEmail": "joanna@163.com",
            "authorUrl": "www.baidu.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "firefox",
            "authorInfo": {
                "userId": "23885819423303763495",
                "fname": "吉",
                "lname": "力",
                "nickName": "机智的码农",
                "avatar": "http://q.qlogo.cn/headimg_dl?dst_uin=806533632&spec=100",
                "birthday": "2017-10-09 17:31:10",
                "sex": "1",
                "siteUrl": "jilijelon.liangrenyixin.cn",
                "introduction": "天下代码，唯快不破"
            },
            "createDate": "2017-12-07 15:00:10",
            "content": "棒棒棒66666",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 1,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24721826429839419921",
            "authorName": "jelon",
            "authorEmail": "jilijelon@163.com",
            "authorUrl": "http://www.baidu.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "fireFox",
            "authorInfo": null,
            "createDate": "2017-12-07 18:49:56",
            "content": "内容也是666",
            "likeCount": 0,
            "replyCount": 3,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 1,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24735712314664979022",
            "authorName": "测试的人",
            "authorEmail": "test@163.com",
            "authorUrl": "www.baidu.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "",
            "authorInfo": null,
            "createDate": "2017-12-08 18:22:28",
            "content": "说点啥好呢",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 0,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24735795014543382943",
            "authorName": "小机智",
            "authorEmail": "xiaojizhi@xiaojizhi.com",
            "authorUrl": "www.asd.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "",
            "authorInfo": null,
            "createDate": "2017-12-08 18:30:53",
            "content": "说点啥好呢",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 0,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24735796165355549736",
            "authorName": "小机智",
            "authorEmail": "xiaojizhi@xiaojizhi.com",
            "authorUrl": "www.asd.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "",
            "authorInfo": null,
            "createDate": "2017-12-08 18:31:00",
            "content": "说点啥好呢2",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 0,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24735796673587217928",
            "authorName": "小机智",
            "authorEmail": "xiaojizhi@xiaojizhi.com",
            "authorUrl": "www.asd.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "",
            "authorInfo": null,
            "createDate": "2017-12-08 18:31:03",
            "content": "说点啥好呢3",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 0,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24735797155276862330",
            "authorName": "小机智",
            "authorEmail": "xiaojizhi@xiaojizhi.com",
            "authorUrl": "www.asd.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "",
            "authorInfo": null,
            "createDate": "2017-12-08 18:31:06",
            "content": "说点啥好呢5",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 0,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }, {
            "id": "24735797635655745425",
            "authorName": "小机智",
            "authorEmail": "xiaojizhi@xiaojizhi.com",
            "authorUrl": "www.asd.com",
            "authorIp": "127.0.0.1",
            "authorAgent": "",
            "authorInfo": null,
            "createDate": "2017-12-08 18:31:09",
            "content": "说点啥好呢5asdas  qwdasd ",
            "likeCount": 0,
            "replyCount": 0,
            "isTop": 0,
            "isHot": 0,
            "type": 0,
            "status": 0,
            "parentId": "0",
            "topicId": "24695531230177356545",
            "topicType": 0
        }]
    });

    console.log($('#my-nest').nestComment('getData'));

});


src/nest-comment.js
/**
 * @author jelon <jilijelon@163.com>
 * version: 1.0.0
 * https://github.com/jilijelon/nest-comment
 */

(function ($) {
    'use strict';

    // TOOLS DEFINITION
    // ======================

    // it only does '%s', and return '' when arguments are undefined
    var sprintf = function (str) {
        var args = arguments,
            flag = true,
            i = 1;

        str = str.replace(/%s/g, function () {
            var arg = args[i++];

            if (typeof arg === 'undefined') {
                flag = false;
                return '';
            }
            return arg;
        });
        return flag ? str : '';
    };

    var calculateObjectValue = function (self, name, args, defaultValue) {
        var func = name;

        if (typeof name === 'string') {
            // support obj.func1.func2
            var names = name.split('.');

            if (names.length > 1) {
                func = window;
                $.each(names, function (i, f) {
                    func = func[f];
                });
            } else {
                func = window[name];
            }
        }
        if (typeof func === 'object') {
            return func;
        }
        if (typeof func === 'function') {
            return func.apply(self, args || []);
        }
        if (!func && typeof name === 'string' && sprintf.apply(this, [name].concat(args))) {
            return sprintf.apply(this, [name].concat(args));
        }
        return defaultValue;
    };

    // NEST COMMENT CLASS DEFINITION
    // ======================
    var NestComment = function (el, options) {
        this.options = options;
        this.$el = $(el);
        this.$el_ = this.$el.clone();

        this.init();
    };

    NestComment.DEFAULTS = {
        classes: 'nest-comment',
        topicType:null,
        topicId:null,
        pagination: true,
        data: [],
        totalField: 'total',
        dataField: 'rows',
        sidePagination:'server',
        totalRows: 0, // server side need to set
        pageNumber: 1,
        pageSize: 10,
        onAll: function (name, args) {
            return false;
        },
        onLoadSuccess: function (data) {
            return false;
        },
        onLoadError: function (status) {
            return false;
        },
        onPageChange: function (number, size) {
            return false;
        },
    };

    NestComment.EVENTS = {
        'all.nest.comment': 'onAll',
        'load-success.nest.comment': 'onLoadSuccess',
        'load-error.nest.comment': 'onLoadError',
        'page-change.nest.comment': 'onPageChange'
    };

    NestComment.prototype.init = function () {
        this.initContainer();
        this.initData();
        this.initCommentBox();
        this.initRespondBox();
        this.initServer();
    }

    // PUBLIC FUNCTION DEFINITION
    // =======================
    NestComment.prototype.initContainer = function () {
        this.$container = this.$el;
        this.$container.addClass(this.options.classes);

        this.$container.append($([
            '<div class="comment-box">',
                '<h3 class="comment-title">',
                '全部评论： <span class="comment-count">0条</span>',
                '</h3>',
                '<div class="comment-loading">',
                '<span><i class="icon-spin6 animate-spin"></i> 加载中...</span>',
                '</div>',
                '<ol class="comment-list">',
                '</ol>',
                '<div class="comment-pagination">',
                '</div>',
            '</div>',
            '<div class="respond-box">'
                '<h3 class="comment-title">发表评论',
                '<span class="comment-reply-cancel"><a rel="nofollow" href="javascript:;">取消回复</a></span> ',
                '</h3>',
                '<form class="comment-form">',
                '<div class="clearfix author-info">',
                '<div class="info-item-3">',
                '<label for="author">昵称<span class="required">*</span></label>',
                '<input type="text" name="author" value="" size="22" tabindex="1" class="comment-md-9 author" />',
                '</div>',
                '<div class="info-item-3">',
                '<label for="email">邮箱<span class="required">*</span></label>',
                '<input type="text" name="email" value="" size="22" tabindex="2" class="comment-md-9 author" />',
                '</div>',
                '<div class="info-item-3">',
                '<label for="url">网址</label>',
                '<input type="text" name="url" value="" size="22" tabindex="3" class="comment-md-9 author" />',
                '</div>',
                '</div>',
                '<div class="comment-from-main clearfix">',
                '<div class="comment-form-textarea">',
                '<div class="comment-textarea-box">',
                '<textarea name="comment" placeholder="说点什么吧..." tabindex="4" class="comment-textarea"></textarea>',
                '</div>',
                '</div>',
                '<div class="form-submit">',
                '<input id="submit" name="submit" type="button" tabindex="5" title="发表评论" value="发表评论" class="btn-comment pull-right" />',
                '<div class="form-data hidden">',
                '</div>',
                '</form>'
            ,'</div>'
        ].join('')));

        this.$commentBoxContainer = this.$container.find('.comment-box');
        this.$respondBoxContainer = this.$container.find('.respond-box');
        this.$commentList = this.$container.find('.comment-list');

        this.$container.after('<div class="clearfix"></div>');
    };

    NestComment.prototype.initCommentBox = function () {
        var data = this.getData();

        for(var i=0;i<data.length;i++){

            this.$commentContainer = $([
                '<li class="depth-1">',
                '<div class="comment">',
                '<div class="comment-avatar">',
                '</div>',
                '<div class="comment-body">',
                '<div class="comment-author"></div>',
                '<div class="comment-reply"></div>',
                '<div class="comment-text"></div>',
                '</div>',
                '</div>',
                '</li>'
            ].join(''));
        }

    };

    NestComment.prototype.initRespondBox = function () {

    };

    /**
     * @param data
     * @param type: append / prepend
     */
    NestComment.prototype.initData = function (data, type) {
        if (type === 'append') {
            this.data = this.data.concat(data);
        } else if (type === 'prepend') {
            this.data = [].concat(data).concat(this.data);
        } else {
            this.data = data || this.options.data;
        }

        // Fix #839 Records deleted when adding new row on filtered table
        if (type === 'append') {
            this.options.data = this.options.data.concat(data);
        } else if (type === 'prepend') {
            this.options.data = [].concat(data).concat(this.options.data);
        } else {
            this.options.data = this.data;
        }

    };

    NestComment.prototype.initServer = function (silent, query, url) {
        var that = this,
            data = {},
            params = {
                //topicType: this.options.topicType,
                topicType:0,
                topicId: '24695531230177356545',
                type:'1',
                page:'1',
                rows:10,
                parentId:0
            },
            request;

        if (!(url || this.options.url) && !this.options.ajax) {
            return;
        }

        data = calculateObjectValue(this.options, this.options.queryParams, [params], data);

        $.extend(data, query || {});

        // false to stop request
        if (data === false) {
            return;
        }

        request = $.extend({}, calculateObjectValue(null, this.options.ajaxOptions), {
            type: this.options.method,
            url: url || this.options.url,
            data: this.options.contentType === 'application/json' && this.options.method === 'post' ?
                JSON.stringify(data) : data,
            cache: this.options.cache,
            contentType: this.options.contentType,
            dataType: this.options.dataType,
            success: function (res) {
                res = calculateObjectValue(that.options, that.options.responseHandler, [res], res);

                that.load(res);
                that.trigger('load-success', res);
                //if (!silent) that.$tableLoading.hide();
            },
            error: function (res) {
                var data = [];

                that.load(data);
                that.trigger('load-error', res.status, res);
                if (!silent) that.$tableLoading.hide();
            }
        });

        if (this.options.ajax) {
            calculateObjectValue(this, this.options.ajax, [request], null);
        } else {
            if (this._xhr && this._xhr.readyState !== 4) {
                this._xhr.abort();
            }
            this._xhr = $.ajax(request);
        }
    };

    NestComment.prototype.initBody = function () {
        var that = this,
            html = [],
            data = this.getData();

        this.trigger('pre-body', data);


        this.trigger('post-body', data);
    };

    NestComment.prototype.getData = function (useCurrentPage) {
        var data = this.options.data;

        return data;
    };

    NestComment.prototype.load = function (data) {

        if (this.options.pagination && this.options.sidePagination === 'server') {
            this.options.totalRows = data[this.options.totalField];
            data = data[this.options.dataField];
        } else if (!$.isArray(data)) {
            data = data.data;
        }

        this.initData(data);

        //this.initPagination();
        this.initBody(fixedScroll);
    };

    NestComment.prototype.trigger = function (name) {
        var args = Array.prototype.slice.call(arguments, 1);

        name += '.nest.comment';
        this.options[NestComment.EVENTS[name]].apply(this.options, args);
        this.$el.trigger($.Event(name), args);

        this.options.onAll(name, args);
        this.$el.trigger($.Event('all.nest.comment'), [name, args]);
    };

    NestComment.prototype.sayHello = function () {
        console.info('hello world');
    };


    // NEST COMMENT PLUGIN DEFINITION
    // =======================
    var allowedMethods = [
        'getOptions',
        'getSelections',
        'getData',
        'sayHello'
    ];

    $.fn.nestComment = function (option) {
        var value,
            args = Array.prototype.slice.call(arguments, 1);

        this.each(function () {
            var $this = $(this),
                data = $this.data('nest.comment'),
                options = $.extend({}, NestComment.DEFAULTS, $this.data(),
                    typeof option === 'object' && option);

            if (typeof option === 'string') {
                if ($.inArray(option, allowedMethods) < 0) {
                    throw new Error("Unknown method: " + option);
                }

                if (!data) {
                    return;
                }

                value = data[option].apply(data, args);

                if (option === 'destroy') {
                    $this.removeData('nest.comment');
                }
            }

            if (!data) {
                $this.data('nest.comment', (data = new NestComment(this, options)));
            }
        });

        return typeof value === 'undefined' ? this : value;
    };

    $.fn.nestComment.Constructor = NestComment;
    $.fn.nestComment.defaults = NestComment.DEFAULTS;
    $.fn.nestComment.methods = allowedMethods;
    /*$.fn.nestComment.utils = {

    };*/



    // NEST COMMENT INIT
    // =======================

    $(function () {
        $('[data-toggle="comment"]').nestComment();
    });
})(jQuery);
